pkg/segment/structs/segstructs.go

```diff
@@ -961,6 +973,201 @@ func (qtype QueryType) String() string {
 	}
 }
 
+func (qa *QueryAggregators) HasStatsBlock() bool {
+	return qa != nil && !qa.HasStreamStats() && qa.HasGroupByOrMeasureAggsInBlock()
+}
+
+func (qa *QueryAggregators) HasStatsBlockInChain() bool {
+	if qa == nil {
+		return false
+	}
+
+	if qa.HasStatsBlock() {
+		return true
+	}
+
+	if qa.Next != nil {
+		return qa.Next.HasStatsBlockInChain()
+	}
+
+	return false
+}
+
+// returns all columns in the query aggregators if stats are present.
+// Update this function whenever a new struct or an is added to the query aggregators.
+func (qa *QueryAggregators) GetAllColsInAggsIfStatsPresent() map[string]struct{} {
+	if qa == nil {
+		return nil
+	}
+
+	if !qa.HasStatsBlockInChain() {
+		return nil
+	}
+
+	cols := make(map[string]struct{})
+
+	qa.GetAllColsInAggsChainUntilStatsBlock(cols)
+
+	return cols
+}
+
+func (qa *QueryAggregators) GetAllColsInAggsChainUntilStatsBlock(cols map[string]struct{}) {
+	if qa == nil {
+		return
+	}
+
+	AddAllColumnsInOutputTransforms(cols, qa.OutputTransforms)
+	AddAllColumnsInMeasureAggs(cols, qa.MeasureOperations)
+	AddAllColumnsInGroupByRequest(cols, qa.GroupByRequest)
+	AddAllColumnsInTransactionArguments(cols, qa.TransactionArguments)
+	AddAllColumnsInStreamStatsOptions(cols, qa.StreamStatsOptions)
+
+	if qa.HasStatsBlock() {
+		// We want to stop at the first stats block
+		return
+	}
+
+	if qa.Next != nil {
+		qa.Next.GetAllColsInAggsChainUntilStatsBlock(cols)
+	}
+}
+
+func AddAllColumnsInOutputTransforms(cols map[string]struct{}, outputTransforms *OutputTransforms) {
+	if outputTransforms == nil {
+		return
+	}
+
+	if outputTransforms.HarcodedCol != nil {
+		sutils.ConvertToSetFromSlice(cols, outputTransforms.HarcodedCol)
+	}
+
+	if outputTransforms.RenameHardcodedColumns != nil {
+		sutils.ConvertToSetFromMap(cols, outputTransforms.RenameHardcodedColumns)
```

Comment:
nit: changing the names to `ConvertSliceToSet` and `ConvertMapToSet` would be slightly more concise. But you can keep what you have if you want